src/function.cpp

```diff
@@ -51,19 +61,27 @@ DeclaredFunction DeclaredFunction::declare(std::vector<mlir::Type> &&domain,
   transform(domain.cbegin(), domain.cend(), back_inserter(smtDomain),
             getScalarSort);
 
-  if (range.isa<mlir::TensorType>()) {
-    throw UnsupportedException(
-        "Function that returns tensor is not supported yet");
-  }
-  if (range.isa<mlir::MemRefType>()) {
+  if (range.isIntOrIndexOrFloat()) {
+    const auto smtRange = getScalarSort(range);
+    FnDecl decl(smtDomain, smtRange, string(name) + "_tvfn");
+    return DeclaredFunction(move(domain), move(range), move(decl));
+  } else if (range.isa<mlir::TensorType>()) {
+    const auto tensorRange = range.dyn_cast<mlir::TensorType>();
+    const auto tensorElementType = getShapeElemType(tensorRange);
+    const auto dims = getShapeDimVector(tensorRange);
+    const auto smtRange =
+        static_cast<Expr>(Tensor::var(tensorElementType,
+                                      string(name) + "_tv_ret_tensor", dims))
+            .sort();
```

Comment:
value.cpp has `arraySortForTensor`.
Let's use it after renaming it to `Tensor::getSort`.